docs/swift/DesignDoc.md

```diff
@@ -137,65 +152,117 @@ lets users treat messages polymorphically. Any shared method implementations
-    nilled out).
-*   Reading the value of a property returns its value if it is set, or returns
-    its default value if it is not set. Reading a property never returns nil.
-
-The final point in the list above implies that the optional cannot be checked to
-determine if the field is set to a value other than its default: it will never
-be nil. Instead, we must provide `has*` methods for each field to allow the user
-to check this. These methods will be public in proto2. In proto3, these methods
-will be private (if generated at all), since the user can test the returned
-value against the zero value for that type.
+The user-facing API for a `ProtoMessage` will be as follows:
+
+```swift
+public protocol ProtoMessage {
+
+  var serializedSize: Int { get }
+
+  init()
+  init(data: NSData) throws
+
+  mutating func merge(fromData data: NSData) throws
+
+  func data() throws -> NSData
+  func write(toOutputStream stream: NSOutputStream) throws
+
+  // Public only as an implementation detail:
+
+  mutating func merge(fromCodedInputStream stream: CodedInputStream) throws
+  func write(toCodedOutputStream stream: CodedOutputStream) throws
+}
+```
+
+The methods listed at the end are public only as an implementation detail,
+because the generated code will be implemented in terms of them to read and
+write the proto wire format. Users should never need to call them directly or
+use the coded stream types. The remaining methods and initializers are
+implemented in a protocol extension in terms of those coded stream methods.
+
+Conformance to `Equatable`, `Hashable`, and `CustomDebugStringConvertible` are
+also desirable but not currently shown, to focus on a small API intended for
+mobile clients. The implementations and generation of those methods are fairly
+straightforward, but it should be noted that generating equality and hashability
+for very large messages that are not likely to use those features would add
+significant bloat to an application. (Mirrors could be investigated as a way of
+reducing this duplication using a visitor pattern, albeit with some performance
+trade-offs.)
+
+### Fields
+
+Fields in a message will be represented as non-optional properties. We explored
+using optionals (regular and implicitly unwrapped), but determined that they
+were unsuitable; protocol buffer users expect to receive default values, not
+`nil`, when retrieving fields, and this is more important in proto2 where fields
+can have explicit non-zero defaults. Implicitly unwrapped computed properties
+could mitigate the `clear` situation by treating assignment of `nil` as clearing
+the field, but we would still need to create a `has` method because the property
+getter would never return nil (it would always return the default value, and
+there would be no way to distinguish "not set" from "explicitly set to the
+default value"). In that sense, having the property be optional seems misleading
+to users.
```

Comment:
I think this discussion is largely recreating the rationale for dropping "required" from proto3:
  * Other languages do not actually require that "required" fields be set
  * Other languages seem to treat "required" and "optional" pretty much the same (some implementations used to warn if a required field was missing, but I believe that practice has faded).

The issue here is that Swift offers a language-level notion of "optional" that seems very awkward when applied to "required" fields.  I haven't found any approach yet that I'm really happy with.

I've tried both of these approaches:
  * Treat "required" fields the same as "optional".  That's essentially what other language implementations have done, but it seems quite confusing to have a "required" field (in the proto) return an Optional<T> (in Swift).
  * In my current implementation, "required" fields are actually required in that they always have a value and are always serialized.  (However, I don't error if they're missing on deserialization; that just leaves the default behind.)  But this seems to scandalize a lot of folks, so I suppose I'll have to change this.

The other options I can see:
  * Treat them like proto3 singular:  They are not Swift optionals, always have a value, but are serialized only if they differ from the default.  This provides a nice clean Swift API but still differs from the C++ implementation, which seems to be a point of concern.
  * Treat them as Swift optionals internally, but not via the API.  They are serialized only if non-nil (internally).

This last option begs the question of whether the user needs to be able to explicitly unset them and/or explicitly test them for being set.  If so, then the most straightforward way to provide that is to expose the field as a Swift optional, which brings us back to the first idea of treating "optional" and "required" as synonyms.

Based on the discussion so far, it seems that treating "optional" as a synonym for "required" is the best match for what other language implementations do.